documentation/private-link-service/private-link-redirect-map.md

```diff
@@ -0,0 +1,256 @@
+| SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_BACKEND_PORT_BASE | sai_uint16_t | Back end port base, aka begin port of translated port range |
+| SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_COUNTER_ID | sai_object_id_t | A counter that tracks the number of packets and bytes associated with the matching port range entry  |
+
+The entry attribute **Action** is defined as below:
+
+```c
+typedef enum _sai_outbound_port_map_port_range_entry_action_t
+{
+    SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ACTION_SKIP_MAPPING,
+    SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ACTION_MAP_TO_PRIVATE_LINK_SERVICE,
+} sai_outbound_port_map_port_range_entry_action_t;
+```
+
+The action `skip_mapping` means to skip the mapping operation. The action `map_to_private_link_service` means to do private link mapping operation as below:
+
+```c
+packet.underlay.destination_ip = attr_BACKEND_IP
+packet.overlay.ipv6.src_addr = ((((bit<128>)packet.overlay.ipv4.src_addr & ~ca_pa.attr_OVERLAY_SIP_MASK) | ca_pa.attr_OVERLAY_SIP) & ~eni.attr_SAI_ENI_ATTR_PL_SIP_MASK) | eni.attr_SAI_ENI_ATTR_PL_SIP
+packet.overlay.ipv6.dst_addr = ((bit<128>)attr_BACKEND_IP & ~ca_pa.attr_OVERLAY_DIP_MASK) | ca_pa.attr_OVERLAY_DIP
+packet.overlay.destination_port = attr_BACKEND_PORT_BASE + (packet.overlay.destination_port - attr_MATCH_PORT_BASE)
+```
+
+### 5.3. DASH CA-PA mapping attributes
+
+The following attributes will be added to CA-to-PA entry, for supporting service rewrites for PL/PL NSG redirect map:
+
+| Attribute name | Type | Description |
+| --- | --- | --- |
+| SAI_OUTBOUND_CA_TO_PA_ENTRY_ATTR_OUTBOUND_PORT_MAP_ID | sai_object_id_t | Outbound port map id. |
+
+### 5.4. DASH ENI counters
+
+The two drop counters for port-map/port-range-entry missing are added to ENI stats as below:
+
+```c
+typedef enum _sai_eni_stat_t
+{
+    ...
+
+    /** DASH ENI OUTBOUND_PORT_MAP_MISS_DROP_PACKETS stat count */
+    SAI_ENI_STAT_OUTBOUND_PORT_MAP_MISS_DROP_PACKETS,
+
+    /** DASH ENI OUTBOUND_PORT_RANGE_ENTRY_MISS_DROP_PACKETS stat count */
+    SAI_ENI_STAT_OUTBOUND_PORT_RANGE_ENTRY_MISS_DROP_PACKETS,
+
+} sai_eni_stat_t;
+```
+
+## 6. DASH pipeline behavior
+
+Following [DASH pipeline behavior of private link service](https://github.com/sonic-net/DASH/blob/main/documentation/private-link-service/private-link-service.md#6-dash-pipeline-behavior), it adds updates for redirect map.
+
+### 6.1  VM-to-PLS direction (Outbound)
+
+**Mapping - VNET**:
+The inner destination IP will be used for finding the outbound CA-PA mapping entry, of which the action will be set to `SAI_OUTBOUND_CA_TO_PA_ENTRY_ACTION_SET_PRIVATE_LINK_MAPPING`. The mapping entry will be associated with the service rewrite info and a port map object:
+
+   | SAI field name | Type | Value |
+   | --- | --- | --- |
+   | entry_attr.outbound_port_map_id | `sai_object_id_t` | (SAI object ID of the port map) |
+   | entry_attr.SAI_OUTBOUND_CA_TO_PA_ENTRY_ATTR_OVERLAY_SIP | `sai_ip_address_t` | `fd40:108:0:5678:0:200::` |
```

Comment:
Hi Junhua, for the example, do you mind to help use a more generic prefix? Like 9999:8888:7777:.... and 1111:2222:3333:4444:....

Reply:
Updated